File: sidecar/test/unit/pick.test.js
'use strict';
const { test } = require('node:test');
const assert = require('node:assert/strict');
const { choosePick } = require('../../pick');

// already ranked weakest-first by select.rankUnits
const RANKED = [
  { path: 'src/main/java/a/JSONArray.java::getNumber', file: 'src/main/java/a/JSONArray.java', method: 'getNumber' },
  { path: 'src/main/java/ui/Widget.java::render', file: 'src/main/java/ui/Widget.java', method: 'render' },
  { path: 'src/main/java/a/JSONObject.java::hashCode', file: 'src/main/java/a/JSONObject.java', method: 'hashCode' },
];

test('with nothing excluded, the pick is the top-ranked unit', () => {
  // the model used to CHOOSE, and chose the 5th-ranked private method over three
  // directly callable ones. Its job is the team rule; the ranking is measurement.
  const r = choosePick(RANKED, []);
  assert.equal(r.file, 'src/main/java/a/JSONArray.java::getNumber');
  assert.match(r.reason, /rank|weakest/i);
});

test('an excluded unit is skipped and the next survivor taken', () => {
  const r = choosePick(RANKED, ['src/main/java/a/JSONArray.java::getNumber']);
  assert.equal(r.file, 'src/main/java/ui/Widget.java::render');
});

test('excluding a FILE excludes every method in it', () => {
  // "don't touch ui" is a statement about files, and the model answers in those terms
  const r = choosePick(RANKED, ['src/main/java/ui/Widget.java', 'src/main/java/a/JSONArray.java']);
  assert.equal(r.file, 'src/main/java/a/JSONObject.java::hashCode');
});

test('a rule that excludes everything ends the batch and is not retried', () => {
  const r = choosePick(RANKED, RANKED.map((c) => c.path));
  assert.equal(r.file, null);
  assert.equal(r.retry, false);
  assert.match(r.reason, /excluded/i);
});

test('an exclusion naming something not on the list is ignored', () => {
  const r = choosePick(RANKED, ['src/main/java/nowhere/Ghost.java::gone']);
  assert.equal(r.file, 'src/main/java/a/JSONArray.java::getNumber');
});

test('no candidates at all is not a rule decision', () => {
  const r = choosePick([], []);
  assert.equal(r.file, null);
  assert.equal(r.retry, false);
  assert.match(r.reason, /no candidates/i);
});

test('the choice is a function of the ranking alone — same input, same unit', () => {
  assert.equal(choosePick(RANKED, []).file, choosePick(RANKED, []).file);
});

File: backend/src/test/java/tech/mikhailov/ijt/PickTest.java
package tech.mikhailov.ijt;

import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.regex.Pattern;

import static org.junit.jupiter.api.Assertions.*;

class PickTest {

    // already ranked weakest-first by select.rankUnits
    private static final List<Pick.Candidate> RANKED = List.of(
            new Pick.Candidate("src/main/java/a/JSONArray.java::getNumber", "src/main/java/a/JSONArray.java", "getNumber"),
            new Pick.Candidate("src/main/java/ui/Widget.java::render", "src/main/java/ui/Widget.java", "render"),
            new Pick.Candidate("src/main/java/a/JSONObject.java::hashCode", "src/main/java/a/JSONObject.java", "hashCode"));

    private static void assertMatches(String regex, String actual) {
        assertNotNull(actual);
        assertTrue(Pattern.compile(regex, Pattern.CASE_INSENSITIVE).matcher(actual).find(),
                actual + " does not match /" + regex + "/i");
    }

    @Test
    void withNothingExcludedThePickIsTheTopRankedUnit() {
        // the model used to CHOOSE, and chose the 5th-ranked private method over three
        // directly callable ones. Its job is the team rule; the ranking is measurement.
        var r = Pick.choosePick(RANKED, List.of());
        assertEquals("src/main/java/a/JSONArray.java::getNumber", r.file());
        assertMatches("rank|weakest", r.reason());
    }

    @Test
    void anExcludedUnitIsSkippedAndTheNextSurvivorTaken() {
        var r = Pick.choosePick(RANKED, List.of("src/main/java/a/JSONArray.java::getNumber"));
        assertEquals("src/main/java/ui/Widget.java::render", r.file());
    }

    @Test
    void excludingAFileExcludesEveryMethodInIt() {
        // "don't touch ui" is a statement about files, and the model answers in those terms
        var r = Pick.choosePick(RANKED, List.of("src/main/java/ui/Widget.java", "src/main/java/a/JSONArray.java"));
        assertEquals("src/main/java/a/JSONObject.java::hashCode", r.file());
    }

    @Test
    void aRuleThatExcludesEverythingEndsTheBatchAndIsNotRetried() {
        var r = Pick.choosePick(RANKED, RANKED.stream().map(Pick.Candidate::path).toList());
        assertNull(r.file());
        assertEquals(Boolean.FALSE, r.retry());
        assertFalse(r.shouldRetry());
        assertMatches("excluded", r.reason());
    }

    @Test
    void anExclusionNamingSomethingNotOnTheListIsIgnored() {
        var r = Pick.choosePick(RANKED, List.of("src/main/java/nowhere/Ghost.java::gone"));
        assertEquals("src/main/java/a/JSONArray.java::getNumber", r.file());
    }

    @Test
    void noCandidatesAtAllIsNotARuleDecision() {
        var r = Pick.choosePick(List.of(), List.of());
        assertNull(r.file());
        assertEquals(Boolean.FALSE, r.retry());
        assertFalse(r.shouldRetry());
        assertMatches("no candidates", r.reason());
    }

    @Test
    void theChoiceIsAFunctionOfTheRankingAloneSameInputSameUnit() {
        assertEquals(Pick.choosePick(RANKED, List.of()).file(), Pick.choosePick(RANKED, List.of()).file());
    }

    @Test
    void aSuccessfulPickStatesNoRetryAndReadsAsFalsy() {
        // Java-side only: the JS answer for a pick has no `retry` key, so the boxed field is
        // null — never true. Unboxing it would throw where the JS merely read undefined, so
        // callers go through shouldRetry().
        var r = Pick.choosePick(RANKED, List.of());
        assertNull(r.retry());
        assertFalse(r.shouldRetry());
    }
}
